api/plgd/plgd_time.c

```diff
@@ -560,4 +598,243 @@ plgd_time_configure(bool use_in_mbedtls,
+{
+  time_fetch_param_t *fp = (time_fetch_param_t *)data->user_data;
+  oc_clock_time_t time = 0;
+  oc_status_t code = data->code;
+  if (code == OC_STATUS_OK &&
+      !dev_time_parse_fetch_response(data->payload, &time)) {
+    code = OC_STATUS_INTERNAL_SERVER_ERROR;
+  }
+
+  OC_DBG("plgd-time: on_fetch time=%d time=%u", (int)code, (unsigned)time);
+  fp->on_fetch(code, time, fp->on_fetch_data);
+#if defined(OC_TCP) || defined(OC_SECURITY)
+  if (fp->close_peer_after_fetch) {
+    OC_DBG("plgd-time: close fetch time session");
+    oc_close_session(data->endpoint);
+  }
+#endif /* OC_TCP || OC_SECURITY */
+
+  oc_memb_free(&g_fetch_params_s, fp);
+}
+
+#if defined(OC_SECURITY) && defined(OC_PKI)
+
+static void
+time_verify_certificate_params_free(void *data)
+{
+  if (data == NULL) {
+    return;
+  }
+  time_verify_certificate_params_t *vcp =
+    (time_verify_certificate_params_t *)data;
+  oc_memb_free(&g_time_verify_certificate_params_s, vcp);
+}
+
+static int
+dev_time_verify_certificate(oc_tls_peer_t *peer, const mbedtls_x509_crt *crt,
+                            int depth, uint32_t *flags)
+{
+  OC_DBG("plgd-time: verifying certificate at depth %d, flags %u", depth,
+         *flags);
+
+  const time_verify_certificate_params_t *vcp =
+    (time_verify_certificate_params_t *)peer->user_data.data;
+  int ret = vcp->verify_certificate(peer, crt, depth, flags);
+  OC_DBG("plgd-time: default verification done (depth=%d, flags=%u)", depth,
+         *flags);
+
+  *flags &=
+    ~((uint32_t)(MBEDTLS_X509_BADCERT_EXPIRED | MBEDTLS_X509_BADCERT_FUTURE));
+  OC_DBG("plgd-time: removed validity errors (flags=%u)", *flags);
+
+  return ret;
+}
+
+static bool
+dev_time_add_peer(const oc_endpoint_t *ep, bool disable_time_verification)
+{
+  OC_DBG("plgd-time: add new peer");
+  time_verify_certificate_params_t *vcp =
+    (time_verify_certificate_params_t *)oc_memb_alloc(
+      &g_time_verify_certificate_params_s);
+  if (vcp == NULL) {
+    OC_ERR("plgd-time add peer failed: cannot allocate verify certificate "
+           "parameters");
+    return false;
+  }
+
+  oc_tls_peer_t *peer = oc_tls_add_peer(ep, MBEDTLS_SSL_IS_CLIENT);
+  if (peer == NULL) {
+    OC_ERR("plgd-time add peer failed: oc_tls_add_peer failed");
+    oc_memb_free(&g_time_verify_certificate_params_s, vcp);
+    return false;
+  }
+
+  if (disable_time_verification) {
+    OC_DBG("plgd-time: disable time verification for peer");
+    vcp->verify_certificate = peer->verify_certificate;
+    peer->user_data.data = vcp;
+    peer->user_data.free = time_verify_certificate_params_free;
+    peer->verify_certificate = dev_time_verify_certificate;
+  }
+
+  return true;
+}
+
+#endif /* OC_SECURITY && OC_PKI */
+
+#if defined(OC_TCP) || defined(OC_SECURITY)
+
+static bool
+dev_time_is_connected(const oc_endpoint_t *endpoint)
+{
+#ifdef OC_SECURITY
+  if ((endpoint->flags & SECURED) != 0) {
+    OC_DBG("plgd-time: peer connected");
+    return oc_tls_connected(endpoint);
```

Comment:
use check if peer exist instead of `oc_tls_connected `
oc_tls_connected  - return true when handshake is over, but if handshake is in progress returns false, so you handle it as not connected  - so `oc_tls_add_peer` returns already exist peer.